panic on kube-router tun devices

### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

We use kube-router to manage pod subnets routing
https://docs.cilium.io/en/stable/gettingstarted/kube-router/#verify-installation
As your docs mention, there are multiple `tun` devices which routes to nodes in another subnet via overlay.
```
$ ip a
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host
       valid_lft forever preferred_lft forever
2: enp3s0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 9000 qdisc fq_codel state UP group default qlen 1000
    link/ether 30:9c:23:9a:d8:85 brd ff:ff:ff:ff:ff:ff
    inet 172.31.100.131/24 brd 172.31.100.255 scope global dynamic enp3s0
       valid_lft 45310sec preferred_lft 45310sec
    inet6 fe80::329c:23ff:fe9a:d885/64 scope link
       valid_lft forever preferred_lft forever
3: docker0: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc noqueue state DOWN group default
    link/ether 02:42:06:d1:06:f3 brd ff:ff:ff:ff:ff:ff
    inet 172.17.0.1/16 brd 172.17.255.255 scope global docker0
       valid_lft forever preferred_lft forever
4: kube-bridge: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 8980 qdisc noqueue state UNKNOWN group default qlen 1000
    link/ether b6:81:40:4c:c8:42 brd ff:ff:ff:ff:ff:ff
    inet6 fe80::b481:40ff:fe4c:c842/64 scope link
       valid_lft forever preferred_lft forever
5: tunl0@NONE: <NOARP> mtu 1480 qdisc noop state DOWN group default qlen 1000
    link/ipip 0.0.0.0 brd 0.0.0.0
6: tun-17230117201@enp3s0: <POINTOPOINT,NOARP,UP,LOWER_UP> mtu 8960 qdisc noqueue state UNKNOWN group default qlen 1000
    link/ipip 172.31.100.131 peer 172.30.117.201
    inet6 fe80::5efe:ac1f:6483/64 scope link
       valid_lft forever preferred_lft forever
7: tun-17230117200@enp3s0: <POINTOPOINT,NOARP,UP,LOWER_UP> mtu 8960 qdisc noqueue state UNKNOWN group default qlen 1000
    link/ipip 172.31.100.131 peer 172.30.117.200
    inet6 fe80::5efe:ac1f:6483/64 scope link
       valid_lft forever preferred_lft forever
8: tun-172311014@enp3s0: <POINTOPOINT,NOARP,UP,LOWER_UP> mtu 8960 qdisc noqueue state UNKNOWN group default qlen 1000
    link/ipip 172.31.100.131 peer 172.31.101.4
    inet6 fe80::5efe:ac1f:6483/64 scope link
       valid_lft forever preferred_lft forever
9: tun-172311015@enp3s0: <POINTOPOINT,NOARP,UP,LOWER_UP> mtu 8960 qdisc noqueue state UNKNOWN group default qlen 1000
    link/ipip 172.31.100.131 peer 172.31.101.5
    inet6 fe80::5efe:ac1f:6483/64 scope link
       valid_lft forever preferred_lft forever
10: cilium_net@cilium_host: <BROADCAST,MULTICAST,NOARP,UP,LOWER_UP> mtu 9000 qdisc noqueue state UP group default qlen 1000
    link/ether 2a:ea:4a:e5:e6:75 brd ff:ff:ff:ff:ff:ff
    inet6 fe80::28ea:4aff:fee5:e675/64 scope link
       valid_lft forever preferred_lft forever
11: cilium_host@cilium_net: <BROADCAST,MULTICAST,NOARP,UP,LOWER_UP> mtu 9000 qdisc noqueue state UP group default qlen 1000
    link/ether 0a:b0:00:8c:7e:98 brd ff:ff:ff:ff:ff:ff
    inet 172.31.84.137/32 scope link cilium_host
       valid_lft forever preferred_lft forever
    inet6 fe80::8b0:ff:fe8c:7e98/64 scope link
       valid_lft forever preferred_lft forever

$ ip r
default via 172.31.100.1 dev enp3s0 proto dhcp src 172.31.100.131 metric 100
172.17.0.0/16 dev docker0 proto kernel scope link src 172.17.0.1 linkdown
172.31.64.0/24 via 172.31.100.21 dev enp3s0 proto 17
172.31.65.0/24 via 172.31.100.19 dev enp3s0 proto 17
172.31.66.0/24 dev tun-17230117201 proto 17 src 172.31.100.131
172.31.67.0/24 dev tun-17230117200 proto 17 src 172.31.100.131
172.31.68.0/24 via 172.31.100.123 dev enp3s0
172.31.70.0/24 dev tun-172311014 proto 17 src 172.31.100.131
172.31.71.0/24 via 172.31.100.31 dev enp3s0 proto 17
172.31.72.0/24 dev tun-172311015 proto 17 src 172.31.100.131
172.31.73.0/24 via 172.31.100.128 dev enp3s0 proto 17
172.31.75.0/24 via 172.31.100.74 dev enp3s0 proto 17
172.31.76.0/24 via 172.31.100.23 dev enp3s0 proto 17
172.31.77.0/24 via 172.31.100.22 dev enp3s0 proto 17
172.31.78.0/24 via 172.31.100.73 dev enp3s0 proto 17
172.31.79.0/24 via 172.31.100.72 dev enp3s0 proto 17
172.31.81.0/24 via 172.31.100.71 dev enp3s0 proto 17
172.31.82.0/24 via 172.31.100.59 dev enp3s0 proto 17
172.31.83.0/24 via 172.31.100.60 dev enp3s0 proto 17
172.31.84.0/24 via 172.31.84.137 dev cilium_host src 172.31.84.137
172.31.84.137 dev cilium_host scope link
172.31.85.0/24 via 172.31.100.121 dev enp3s0
172.31.86.0/24 via 172.31.100.32 dev enp3s0 proto 17
172.31.87.0/24 via 172.31.100.29 dev enp3s0
172.31.88.0/24 via 172.31.100.122 dev enp3s0
172.31.89.0/24 via 172.31.100.134 dev enp3s0
172.31.92.0/24 via 172.31.100.133 dev enp3s0
172.31.100.0/24 dev enp3s0 proto kernel scope link src 172.31.100.131
172.31.100.1 dev enp3s0 proto dhcp scope link src 172.31.100.131 metric 100
```
Now if we try to start cilium with
```
tunnel: "disabled"
ipam: "kubernetes"
enable-host-legacy-routing: "false"
```
it fails with:
```
...
level=info msg="Inheriting MTU from external network interface" device=enp3s0 ipAddr=172.31.100.131 mtu=9000 subsys=mtu
...
level=info msg="Direct routing device detected" direct-routing-device=enp3s0 subsys=daemon
level=info msg="Detected devices" devices="[enp3s0 tun-17230117200 tun-17230117201 tun-172311014 tun-172311015]" subsys=daemon
...
level=info msg="Initializing ClusterMesh routing" path=/var/lib/cilium/clustermesh/ subsys=daemon
panic: runtime error: index out of range [4] with length 4

goroutine 1 [running]:
github.com/cilium/cilium/pkg/mac.CArrayString({0xc0017760f0, 0xc000c8c4c0, 0x2})
	/go/src/github.com/cilium/cilium/pkg/mac/mac.go:136 +0x185
github.com/cilium/cilium/pkg/datapath/linux/config.devMacros()
	/go/src/github.com/cilium/cilium/pkg/datapath/linux/config/config.go:672 +0x674
github.com/cilium/cilium/pkg/datapath/linux/config.(*HeaderfileWriter).WriteNodeConfig(0xc000052228, {0x2c06f80, 0xc00000eb68}, 0xc00065d088)
	/go/src/github.com/cilium/cilium/pkg/datapath/linux/config/config.go:393 +0x4d65
github.com/cilium/cilium/daemon/cmd.(*Daemon).createNodeConfigHeaderfile(0xc000577b00)
	/go/src/github.com/cilium/cilium/daemon/cmd/datapath.go:61 +0x356
github.com/cilium/cilium/daemon/cmd.(*Daemon).init(0xc000577b00)
	/go/src/github.com/cilium/cilium/daemon/cmd/daemon.go:223 +0x598
github.com/cilium/cilium/daemon/cmd.NewDaemon({0x2c52a70, 0xc0006b0400}, 0xc0008246e0, 0xc00030e150, {0x2cb2da8, 0xc00030e2a0})
	/go/src/github.com/cilium/cilium/daemon/cmd/daemon.go:884 +0x39b2
github.com/cilium/cilium/daemon/cmd.runDaemon()
	/go/src/github.com/cilium/cilium/daemon/cmd/daemon_main.go:1647 +0xb07
github.com/cilium/cilium/daemon/cmd.glob..func1(0x42ee340, {0x27d674c, 0x1, 0x1})
	/go/src/github.com/cilium/cilium/daemon/cmd/daemon_main.go:126 +0x3ca
github.com/spf13/cobra.(*Command).execute(0x42ee340, {0xc00004c030, 0x1, 0x1})
	/go/src/github.com/cilium/cilium/vendor/github.com/spf13/cobra/command.go:860 +0x5f8
github.com/spf13/cobra.(*Command).ExecuteC(0x42ee340)
	/go/src/github.com/cilium/cilium/vendor/github.com/spf13/cobra/command.go:974 +0x3bc
github.com/spf13/cobra.(*Command).Execute(...)
	/go/src/github.com/cilium/cilium/vendor/github.com/spf13/cobra/command.go:902
github.com/cilium/cilium/daemon/cmd.Execute()
	/go/src/github.com/cilium/cilium/daemon/cmd/daemon_main.go:140 +0x45
main.main()
	/go/src/github.com/cilium/cilium/daemon/main.go:16 +0x17
``` 
Even enabling `debug` do not provide any additional info. I thought that issue is somewhere near `WriteNodeConfig` and checked all the node object fields. But it turns out that we can start cilium via excluding `tun` devices, i.e:
```
devices: enp3s+
```
In this case panic on start is gone, and we see such lines in logs:
```
level=info msg="Inheriting MTU from external network interface" device=enp3s0 ipAddr=172.31.100.131 mtu=9000 subsys=mtu
level=info msg="Direct routing device detected" direct-routing-device=enp3s0 subsys=daemon
level=info msg="Detected devices" devices="[enp3s0]" subsys=daemon
```

 Unfortunately when packet arrives via tunnel from pod on another node, the answer is sent to `enp3s0`, not back to tunnel. 
 So, host routing table is ignored. It seems it shouldn't be so for eBPF Host-Routing:
 https://cilium.io/blog/2021/05/11/cni-benchmark#ebpfhostrouting
 > The routing table is consulted directly from eBPF
 
 For now we have to fix broken tunnels like this:
```
enable-host-legacy-routing: "true"
```
This way host routing table is back in business and answer is sent back to tunnel.
Can we have eBPF Host-Routing with kube-router tunnels?

### Cilium Version

cilium-cli: v0.10.2 compiled with go1.17.6 on darwin/amd64
cilium image (default): v1.11.1
cilium image (stable): v1.11.1
cilium image (running): v1.11.1

### Kernel Version

Linux 5.13.0-28-generic #31~20.04.1-Ubuntu SMP Wed Jan 19 14:08:10 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux

### Kubernetes Version

v1.20.11

### Sysdump

Cannot post publicly, but can provide in private.

### Relevant log output

_No response_

### Anything else?

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

--- a/pkg/datapath/linux/config/config.go
+++ b/pkg/datapath/linux/config/config.go
@@ -703,7 +703,7 @@ func devMacros() (string, string, error) {
 		}
 		idx := link.Attrs().Index
 		m := link.Attrs().HardwareAddr
-		if m == nil {
+		if m == nil || len(m) != 6 {
 			l3DevIfIndices = append(l3DevIfIndices, idx)
 		}
 		macByIfIndex[idx] = mac.CArrayString(m)

--- a/pkg/mac/mac.go
+++ b/pkg/mac/mac.go
@@ -128,7 +128,7 @@ func HaveMACAddrs(ifaces []string) bool {
 // CArrayString returns a string which can be used for assigning the given
 // MAC addr to "union macaddr" in C.
 func CArrayString(m net.HardwareAddr) string {
-	if m == nil || len(m) == 0 {
+	if m == nil || len(m) != 6 {
 		return "{0x0,0x0,0x0,0x0,0x0,0x0}"
 	}
 

